samples/order/processor/processor.go

```diff
@@ -0,0 +1,235 @@
+	}
+	if *responseTopicURL == "" {
+		log.Fatal("missing -response-topic")
+	}
+	if *bucketURL == "" {
+		log.Fatal("missing -bucket")
+	}
+	if *collectionURL == "" {
+		log.Fatal("missing -collection")
+	}
+	ctx := context.Background()
+	p, err := newProcessor(ctx, *requestSubURL, *responseTopicURL, *bucketURL, *collectionURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if err := p.run(ctx); err != nil {
+		log.Fatal(err)
+	}
+}
+
+type processor struct {
+	requestSub    *pubsub.Subscription
+	responseTopic *pubsub.Topic
+	bucket        *blob.Bucket
+	coll          *docstore.Collection
+}
+
+func newProcessor(ctx context.Context, subURL, topicURL, bucketURL, collURL string) (*processor, error) {
+	reqSub, err := pubsub.OpenSubscription(ctx, subURL)
+	if err != nil {
+		return nil, err
+	}
+	resTopic, err := pubsub.OpenTopic(ctx, topicURL)
+	if err != nil {
+		return nil, err
+	}
+	bucket, err := blob.OpenBucket(ctx, bucketURL)
+	if err != nil {
+		return nil, err
+	}
+	coll, err := docstore.OpenCollection(ctx, collURL)
+	if err != nil {
+		return nil, err
+	}
+	return &processor{
+		requestSub:    reqSub,
+		responseTopic: resTopic,
+		bucket:        bucket,
+		coll:          coll,
+	}, nil
+}
+
+func (p *processor) run(ctx context.Context) error {
+	for {
+		var req common.OrderRequest
+		msg, err := receiveJSON(ctx, p.requestSub, &req)
+		if err != nil {
+			return err
+		}
+		log.Printf("received %+v", req)
+		res := p.handleOrder(ctx, &req)
+		if res == nil {
+			log.Printf("duplicate finished order %v", req.ID)
+			continue
+		}
+		if err := sendJSON(ctx, p.responseTopic, res); err != nil {
+			return err
+		}
+		log.Printf("sent %+v", res)
+		msg.Ack()
+	}
+}
+
+// handleOrder processes the order request. A processing error is a kind of response.
+func (p *processor) handleOrder(ctx context.Context, req *common.OrderRequest) *common.OrderResponse {
+	res, err := p.processOrder(ctx, req)
+	if err != nil {
+		// TODO(jba): record error metric
+		res = &common.OrderResponse{
+			ID:   req.ID,
+			Note: fmt.Sprintf("processing failed: %v", err),
+		}
+	}
+	return res
+}
+
+func (p *processor) processOrder(ctx context.Context, req *common.OrderRequest) (res *common.OrderResponse, err error) {
+	// See if there is already a document for this order.
+	order := &common.Order{ID: req.ID}
+	err = p.coll.Get(ctx, order)
+	switch {
+	case gcerrors.Code(err) == gcerrors.NotFound:
+		// Normal case: the order hasn't been created yet. Do so.
+		order = &common.Order{
+			ID:         req.ID,
+			Email:      req.Email,
+			InImage:    req.InImage,
+			OutImage:   fmt.Sprintf("%s-out.png", req.InImage),
+			CreateTime: req.CreateTime,
+		}
+		if err := p.coll.Create(ctx, order); err != nil {
+			return nil, err
+		}
+
+	case err == nil && order.FinishTime.IsZero():
+		// The order exists, but was not finished. Either it was abandoned by the processor that
+		// was working on it (probably because the processor died), or it is in progress. Assume
```

Comment:
This part isn't clear to me. You seem to be Acking right after receiving, so the only case in which pubsub will resend is when the crash happened before acking? Or Ack got lost, I guess. But what happens if the processor sends Ack and then doesn't  actually finish processing the order?

Reply:
I ack after receive if there was an error unmarshaling the JSON. On the happy path, I ack only after processing is done.